plugin/pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -1480,3 +1480,49 @@ func (c *VolumeNodeChecker) predicate(pod *v1.Pod, meta algorithm.PredicateMetad
 	}
 	return true, nil, nil
 }
+
+type NoDeletedPVCChecker struct {
+	pvcInfo PersistentVolumeClaimInfo
+}
+
+// NewNoDeletedPVCsPredicate evaluates if a pod can fit due to the PVC it uses.
+func NewNoDeletedPVCsPredicate(pvcInfo PersistentVolumeClaimInfo) algorithm.FitPredicate {
+	c := &NoDeletedPVCChecker{
+		pvcInfo: pvcInfo,
+	}
+	return c.predicate
+}
+
+func (c *NoDeletedPVCChecker) predicate(pod *v1.Pod, meta algorithm.PredicateMetadata, nodeInfo *schedulercache.NodeInfo) (bool, []algorithm.PredicateFailureReason, error) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.PVCProtection) {
+		return true, nil, nil
+	}
+
+	// If a pod doesn't have any volume attached to it, the predicate will always be true.
+	// Thus we make a fast path for it, to avoid unnecessary computations in this case.
+	if len(pod.Spec.Volumes) == 0 {
+		return true, nil, nil
+	}
+
+	namespace := pod.Namespace
+	manifest := &(pod.Spec)
+	for i := range manifest.Volumes {
+		volume := &manifest.Volumes[i]
+		if volume.PersistentVolumeClaim == nil {
+			continue
+		}
+		pvcName := volume.PersistentVolumeClaim.ClaimName
+		if pvcName == "" {
+			return false, nil, fmt.Errorf("PersistentVolumeClaim had no name")
+		}
+		pvc, err := c.pvcInfo.GetPersistentVolumeClaimInfo(namespace, pvcName)
+		if err != nil {
+			return false, nil, err
+		}
+
+		if pvc.DeletionTimestamp != nil {
+			return false, nil, fmt.Errorf("PVC %s is being deleted", pvc.Name)
```

Comment:
If you want to keep this as a predicate, failure reason must be set and error must be nil. The failure reason should be added to the `nodesWherePreemptionMightHelp()` function of generic_scheduler.go as an an unresolvable error.

Another option (recommended one) is to make a function in generic_scheduler.go and call it [here](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/core/generic_scheduler.go#L101). The function can be called something like `podPassesBasicChecks`. If the function returns an error, `Schedule()` function of the generic_scheduler.go returns `"", err` where `err` is the same error as the one returned by `podPassesBasicChecks`. This is a better approach. It won't need changes to other places (such as  `nodesWherePreemptionMightHelp()`) and the check is run only once per pod, not for each node. Future such checks can be added to this newly added function.

Reply:
preempting other pods will not make this pod schedulable.

Reply:
Yes, that's why this error should be added to `nodesWherePreemptionMightHelp()` as an unresolvable error. Unresolvable errors are those scheduling errors that cannot be addressed by preemption. Well, all of this is needed only if we want to keep this as a predicate.

Reply:
> make a function in generic_scheduler.go

Done as you suggested. Note that I needed to pass `pvcLister` into `GenericScheduler` and all tests, therefore the PR has grown a bit.

No alpha feature gate!